python/cornserve/services/resource_manager/manager.py

```diff
@@ -291,7 +291,27 @@ async def scale_up_unit_task(self, task: UnitTask, num_gpus: int) -> None:
                 return
 
             # TODO: decide GPU placement strategy & preference
-            resources = self.resource.allocate(num_gpus=num_gpus, owner=task_state.id)
+            resources = []
+            gpus_to_allocate = num_gpus
+            for chunk_size in sorted(profile.keys(), reverse=True):
+                num_chunks = gpus_to_allocate // chunk_size
+                if num_chunks == 0:
+                    continue
+                for _ in range(num_chunks):
+                    batched_resources = self.resource.allocate(
+                        num_gpus=chunk_size,
+                        owner=task_state.id,
+                    )
+                    resources.extend(batched_resources)
+                gpus_to_allocate %= chunk_size
```

Comment:
```suggestion
                num_chunks, gpus_to_allocate = divmod(gpus_to_allocate, chunk_size)
                if not num_chunks:
                    continue
                for _ in range(num_chunks):
                    batched_resources = self.resource.allocate(
                        num_gpus=chunk_size,
                        owner=task_state.id,
                    )
                    resources.extend(batched_resources)
```
Just made it a bit more Pythonic. But I basically wanted to ask what the concrete case you wanted to quickly fix here. Ideally, it should have been one `resource.allocate` call with `num_gpus` that handles this behavior natively, and doing this loop is more of a hack.

Reply:
I think that's better. I'll move it there

Reply:
After some more careful thought, I stand by what I have here.

What I wanted to fix here is the scale-up endpoint. Without this update, GPU allocation during scaling up is always done one by one, so adding two more GPUs to a unit task is most likely to result in two GPUs on two different nodes, with `node_selection_policy="spread"`. This, however, is unsuitable for the newly added profiles, where the minimal GPU requirement for a task executor could be greater than 1.

My arguments for the current approach are as follows:

1. `resource.allocate` supports the `must_colocate` abstraction, which will be broken in your proposal. For example, allocating `7` GPUs to `2` nodes each with `4` GPUs will inevitably result in `[4,3]`, contradicting `must_colocate`.
2. `resource.allocate` is already a fairly complex function. Adding profile-based allocation logic will inevitably make this function more complex.

I'm inclined to consider `resource.allocate` as a self-contained, unified unit of logic that's used for various scaling approaches. That said, I'm okay with adding another api in `resource.py` to support scaling up (migrate the above logic there).

Reply:
That makes sense. `malloc` has its way of allocating and optimized applications control how `malloc` is called based on application memory usage characteristics. `malloc` doesn't have to have all of that built in.